BUG: Raise correct errors in boolean indexing fast path

asmeurer opened the pull request:
Previously the logic assumed that an index was valid if the size was the same
as the indexed array, rather than the shape. This resulted in an index being
incorrectly allowed if the index was all False, like np.empty((3, 3))[np.full((1, 9),
False)], or incorrectly giving a ValueError about broadcasting if the index
was not all False (like np.empty((3, 3))[np.full((1, 9), True)]). Now these
examples both give IndexError with the usual error message.

Fixes #16997.

(not really sure about the NumPy commit message flags, or which one should apply to this. Is this considered an API break?)

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

asmeurer committed:
Remove incorrect logic from boolean indexing fast path

Previously the logic assumed that an index was valid if the size was the same
as the indexed array, rather than the shape. This resulted in an index being
incorrectly allowed if the index was all False, like np.empty((3, 3))[np.full((1, 9),
False)], or incorrectly giving a ValueError about broadcasting if the index
was not all False (like np.empty((3, 3))[np.full((1, 9), True)]). Now these
examples both give IndexError with the usual error message.

Fixes #16997.

seberg commented:
Probably wouldn't use API, considering how small the change is. But a compatibility release note is necessary, at least for the wrong result → error change.  The comment should now be changed to say that the shape check is always an error, and only there to ensure the same error message as in the generic path (the generic path sets it a bit later, because it may have to expand Ellipsis, etc. so it does not know at that point which dimension the index acts on).

asmeurer committed:
Add another test case for the fixed masking bug

This test used to give a different broadcasting ValueError than the one from
other test.

asmeurer committed:
Update the comments around boolean arrays in mapping.c

asmeurer commented:
I'm still a little shaky on the logic in that function, so you'll have to tell me if my updated comments are correct.

seberg commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -539,22 +539,23 @@ prepare_index(PyArrayObject *self, PyObject *index,
-                 * The reason for that is that fancy indexing uses nonzero,
-                 * and only the result of nonzero is checked for legality.
+                 * If shapes match exactly, this can be optimized as a single
+                 * boolean index. The reason for that is that fancy indexing
+                 * uses nonzero, and only the result of nonzero is checked for
+                 * legality. Otherwise, if the shapes don't match and it isn't
+                 * the scalar case checked below, it should be an error.
```
The nonzero reference is not useful here, and I doubt its correct. Historically, it *was* correct, but that is a different issue. Maybe:

```suggestion
                 * boolean index. When the dimensions are identical but the shapes are not,
                 * this is always an error. The check ensures that these errors are raised
                 * and match those of the generic path.
```

asmeurer committed:
Add a changelog entry for PR #17010

asmeurer commented:
I added a changelog entry. I can't figure out how to build the docs (git submodule is doing something stupid), so I hope it renders OK.

asmeurer committed:
Update a comment according to @seberg's suggestion

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

seberg commented:
@asmeurer no need, it seems fine, and most errors should be noted. You can actually press the `ci/circleci: build artifact` link and go to the page, which is: https://15017-908607-gh.circle-artifacts.com/0/doc/build/html/release/1.20.0-notes.html?highlight=release%20notes#boolean-array-indices-with-mismatching-shapes-now-properly-give-indexerror

Looks good to me. A comment for the tests might be nice, they are hard to understand why the different `indx1`, etc. are all wrong in slightly different ways.  The last one is broadcastable but shape mismatch?

For anyone interested: I am planning to put this in.  It does change a corner case of an all False boolean index of matching size but not shape into an error.  It also changes some ValueError → IndexError, but I think we have been pretty liberal about such changes in the past.

asmeurer commented:
> The last one is broadcastable but shape mismatch?

I guess that's what the difference is. I just found that it gives a different error, `ValueError: non-broadcastable operand with shape (1,1,2) doesn't match the broadcast shape (1,2,2)`, so it presumably goes through a different code path.

asmeurer committed:
Add comments explaining the indexing fast path tests

seberg commented:
Thanks @asmeurer, also for adding the final comments to the tests!
